update/sync/syncHeadersByHash.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/storage"
+	"github.com/ElrondNetwork/elrond-go/update"
+)
+
+type missingHeadersByHash struct {
+	mutMissingHdrs sync.Mutex
+	mapHeaders     map[string]data.HeaderHandler
+	mapHashes      map[string]struct{}
+	pool           dataRetriever.HeadersPool
+	storage        update.HistoryStorer
+	chReceivedAll  chan bool
+	marshalizer    marshal.Marshalizer
+	stopSyncing    bool
+	epochToSync    uint32
+	syncedAll      bool
+	requestHandler process.RequestHandler
+}
+
+// ArgsNewMissingHeadersByHashSyncer defines the arguments needed for the sycner
+type ArgsNewMissingHeadersByHashSyncer struct {
+	Storage        storage.Storer
+	Cache          dataRetriever.HeadersPool
+	Marshalizer    marshal.Marshalizer
+	RequestHandler process.RequestHandler
+}
+
+// NewMissingheadersByHashSyncer creates a syncer for all missing headers
+func NewMissingheadersByHashSyncer(args ArgsNewMissingHeadersByHashSyncer) (*missingHeadersByHash, error) {
+	if check.IfNil(args.Storage) {
+		return nil, dataRetriever.ErrNilHeadersStorage
+	}
+	if check.IfNil(args.Cache) {
+		return nil, update.ErrNilCacher
+	}
+	if check.IfNil(args.Marshalizer) {
+		return nil, dataRetriever.ErrNilMarshalizer
+	}
+	if check.IfNil(args.RequestHandler) {
+		return nil, process.ErrNilRequestHandler
+	}
+
+	p := &missingHeadersByHash{
+		mutMissingHdrs: sync.Mutex{},
+		mapHeaders:     make(map[string]data.HeaderHandler),
+		mapHashes:      make(map[string]struct{}),
+		pool:           args.Cache,
+		storage:        args.Storage,
+		chReceivedAll:  make(chan bool),
+		requestHandler: args.RequestHandler,
+		stopSyncing:    true,
+		syncedAll:      false,
+		marshalizer:    args.Marshalizer,
+	}
+
+	p.pool.RegisterHandler(p.receivedHeader)
+
+	return p, nil
+}
+
+// SyncMissingHeadersByHash syncs the missing headers
+func (m *missingHeadersByHash) SyncMissingHeadersByHash(
+	shardIDs []uint32,
+	headersHashes [][]byte,
+	waitTime time.Duration,
+) error {
+	_ = process.EmptyChannel(m.chReceivedAll)
+
+	requestedMBs := 0
+	m.mutMissingHdrs.Lock()
+	m.stopSyncing = false
+	for index, hash := range headersHashes {
+		m.mapHashes[string(hash)] = struct{}{}
+		header, ok := m.getHeaderFromPoolOrStorage(hash)
+		if ok {
+			m.mapHeaders[string(hash)] = header
+			continue
+		}
+
+		requestedMBs++
+		m.requestHandler.RequestShardHeader(shardIDs[index], hash)
+	}
+	m.mutMissingHdrs.Unlock()
+
+	var err error
+	defer func() {
+		m.mutMissingHdrs.Lock()
+		m.stopSyncing = true
+		if err == nil {
+			m.syncedAll = true
+		}
+		m.mutMissingHdrs.Unlock()
+	}()
+
+	if requestedMBs > 0 {
+		err = WaitFor(m.chReceivedAll, waitTime)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// receivedHeader is a callback function when a new header was received
+// it will further ask for missing transactions
+func (m *missingHeadersByHash) receivedHeader(hdrHandler data.HeaderHandler, hdrHash []byte) {
+	m.mutMissingHdrs.Lock()
```

Comment:
Can you split this function into two? One that implements the critical section and uses the Unlock on defer pattern, and the second to deal with the fake write into the channel?

Also, can you perhaps use a wait group instead of faking waits using a channel? `Add(1)` instead of `EmptyChannel`, `Wait` instead of `WaitFor`, `Done` instead of `<- true`? An alternative to channels, perhaps a bit easier to read. Optional, of course.

Reply:
channels and wait is the current design pattern accross the protocol to request and sync missing stuff.